internal/credentials/hcp/service_principal_test.go

```diff
@@ -0,0 +1,163 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: BUSL-1.1
+
+package hcp
+
+import (
+	"context"
+	"fmt"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/client/fake"
+
+	secretsv1beta1 "github.com/hashicorp/vault-secrets-operator/api/v1beta1"
+	"github.com/hashicorp/vault-secrets-operator/internal/credentials/errors"
+)
+
+func TestServicePrincipleCredentialProvider_GetCreds(t *testing.T) {
+	ctx := context.Background()
+	authObj := &secretsv1beta1.HCPAuth{
+		Spec: secretsv1beta1.HCPAuthSpec{
+			Method: ProviderMethodServicePrincipal,
+			ServicePrincipal: &secretsv1beta1.HCPAuthServicePrincipal{
+				SecretRef: "foo",
+			},
+		},
+	}
+
+	tests := []struct {
+		name              string
+		authObj           *secretsv1beta1.HCPAuth
+		client            client.Client
+		secretData        map[string][]byte
+		providerNamespace string
+		want              map[string]any
+		wantErr           assert.ErrorAssertionFunc
+	}{
+		{
+			name:              "valid",
+			authObj:           authObj,
+			providerNamespace: "tenant-ns",
+			client:            fake.NewClientBuilder().Build(),
+			secretData: map[string][]byte{
+				ProviderSecretClientID:  []byte("client-id-1"),
+				ProviderSecretClientKey: []byte("client-key-1"),
+			},
+			want: map[string]any{
+				ProviderSecretClientID:  "client-id-1",
+				ProviderSecretClientKey: "client-key-1",
+			},
+			wantErr: assert.NoError,
+		},
+		{
+			name:              "secret-not-found",
+			authObj:           authObj,
+			providerNamespace: "tenant-ns",
+			client:            fake.NewClientBuilder().Build(),
```

Comment:
Looks like `client` is defined the same way in all the cases, so we might not need it as a test parameter?

Reply:
Each test case needs its own client instance to ensure it does not pick up any cruft from the other tests.

Reply:
Sure, wasn't suggesting using the same client for all tests, just that the client could be defined at the start of the test run like the other local variables there.

Reply:
Updated with 865c24f050833a8cf29874f2d3003f254d2e6b75